database/migrations/default/1742417174000_crash_ems_link/up.sql

```diff
@@ -0,0 +1,391 @@
+create index ems__incidents_apd_incident_numbers_index on ems__incidents using gin (
+    apd_incident_numbers
+);
+create index ems__incidents_geometry_index on ems__incidents using gist (
+    geometry
+);
+create index ems__incidents_patient_injry_sev_id_index on ems__incidents (
+    patient_injry_sev_id
+);
+create index ems__incidents_crash_match_status_index on ems__incidents (
+    crash_match_status
+);
+create index crashes_case_id_index on crashes (case_id);
+create index crashes_is_temp_record_index on crashes (
+    is_temp_record
+);
+-- these indexes optimize our ST_DWithin operations, which require geography types and are the fastes way to check point-to-point proximity
+create index idx_ems_geometry_geography on ems__incidents using gist (
+    (geometry::geography)
+);
+create index idx_crashes_position_geography on crashes using gist (
+    (position::geography)
+);
+
+--
+-- Create or replace the trigger function for the crashes table
+--
+create or replace function update_crash_ems_match()
+returns trigger as $$
+DECLARE
+    matching_ems RECORD;
+    ems_record RECORD;
+    match_count INTEGER;
+    matched_crash_ids INTEGER[];
+    meters_threshold INTEGER := 1000;
+    time_threshold INTERVAL := '60 minutes';
+BEGIN
+    -- Find all EMS records near the crash location + time
+    FOR matching_ems IN (
+        SELECT 
+            e.id,
+            e.incident_received_datetime,
+            e.geometry
+        FROM 
+            ems__incidents e
+        WHERE
+            -- ignore manual qa matches: these should not be modified
+            e.crash_match_status != 'matched_by_manual_qa'
+            AND e.incident_received_datetime  >= (NEW.crash_timestamp - time_threshold)
+            AND e.incident_received_datetime  <= (NEW.crash_timestamp + time_threshold)
+            AND e.geometry IS NOT NULL
+            AND NEW.position IS NOT NULL
+            -- todo: try use_spheroid = false
+            AND ST_DWithin(e.geometry::geography, NEW.position::geography, meters_threshold)
+    ) LOOP
+        -- Find all crashes which match this EMS record location + time
+        SELECT 
+            ARRAY(
+                SELECT c.id
+                FROM crashes c
+                WHERE 
+                    c.is_deleted = false
+                    and not c.is_temp_record
+                    AND matching_ems.incident_received_datetime  >= (c.crash_timestamp - time_threshold)
+                    AND matching_ems.incident_received_datetime  <= (c.crash_timestamp + time_threshold)
+                    AND c.position IS NOT NULL
+                    AND ST_DWithin(matching_ems.geometry::geography, c.position::geography, meters_threshold)
+            ) INTO matched_crash_ids;
+            
+        -- Get the count from the array length (handling when array_length is null as 0)
+        SELECT COALESCE(array_length(matched_crash_ids, 1), 0) INTO match_count;
+        
+        IF match_count = 0 THEN
```

Comment:
im having trouble wrapping my head around this whole loop part - how would the match count be 0 if we already know the EMS records we are looping thru can at least be linked to the crash that triggered this?

Reply:
ah - i guess if the crash is a temp record is one example. but then why not move that logic further up into what crashes can even trigger this trigger function at all

Reply:
you're right, the only time `match_count` would = 0 would be if a record is a temp record is deleted. 

 was a little hesitant to remove it in the interest of future-proofing. e.g. if in the future we decide to allow, for example, temp records to be changed to permanent records, or we enable soft deleting of non-temp records. i guess if we made such a drastic change we would hopefully search our db schema for references to those columns?